src/FwdState.cc

```diff
@@ -896,12 +906,87 @@ FwdState::noteConnection(HappyConnOpener::Answer &answer)
         if (originWantsEncryptedTraffic && // the "encrypted traffic" part
                 !peer->options.originserver && // the "through a proxy" part
                 !peer->secure.encryptTransport) // the "exclude HTTPS proxies" part
-            return advanceDestination("establish tunnel through proxy", answer.conn, [this,&answer] {
-            establishTunnelThruProxy(answer.conn);
+            return advanceDestination("establish tunnel through proxy", conn, [this,&conn] {
+            establishTunnelThruProxy(conn);
         });
     }
 
-    secureConnectionToPeerIfNeeded(answer.conn);
+    secureConnectionToPeerIfNeeded(conn);
+}
+
+/// whether it is required to send Proxy protocol header
+bool
+FwdState::needProxyProtoHeader() const
+{
+    if (!Config.outgoingProxyProtocolHttp)
+        return false;
+
+    if (!Config.outgoingProxyProtocolHttp->aclList)
+        return true;
+
+    ACLFilledChecklist ch(Config.outgoingProxyProtocolHttp->aclList, request);
+    ch.al = al;
+    ch.syncAle(request, nullptr);
+    return ch.fastCheck().allowed();
+}
+
+void
+FwdState::sendProxyProtoHeaderIfNeeded(const Comm::ConnectionPointer &conn)
+{
+    if (needProxyProtoHeader()) {
+        static const SBuf v2("2.0");
+        ProxyProtocol::Header header(v2, ProxyProtocol::Two::cmdProxy);
+        header.sourceAddress = conn->local; // TODO: must be original source
+        header.destinationAddress = conn->remote;
+
+        BinaryPacker packer;
+        header.pack(packer);
+        const auto packed = packer.packed();
```

Comment:
When you update this code to generate the PROXY protocol  header _before_ opening a connection, please store it as SBuf. Convert to MemBuf only when it is time to write it (using the freshly opened connection).